### PR TITLE
Orca image ImageNet

### PR DESCRIPTION
**Pull request by Le-Zheng**
Download raw ImageNet data “ILSVRC2012_img_train.tar” and “ILSVRC2012_img_val.tar”, extract and provide the raw ImageNet data dir with the format: 
      - train/n03062245/n03062245_4620.JPEG
      - validation/ILSVRC2012_val_00000001.JPEG
The train and validation data will be converted into 1024 and 128 TFRecord files, respectively.
  train/train-00000-of-01024
  ...
  train/train-001023-of-01024
and
  validation/validation-00000-of-00128
  ...
  validation/validation-00127-of-00128
Each train TFRecord file contains ~1250 records. Each validation TFRecord file contains ~390 records. 

Usage:
```
from zoo.orca.data.image.imagenet_dataset import write_imagenet, read_imagenet

write_imagenet(raw_imagenet_path, output_path)
read_imagenet(imagenet_train_path, is_training)
```

`read_imagenet` creates `dataset = tf.data.Dataset.from_tensor_slices(lists(file_path))`  and converts to `dataset = dataset.interleave(tf.data.TFRecordDataset)`

### CODE REVIEW

File: pyzoo/zoo/orca/data/image/imagenet_dataset.py [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+            - Training images: train/train-00000-of-01024
+            - Validation Images: validation/validation-00000-of-00128
+    is_training: True or False
+
+    """
+    filenames = get_filenames(is_training, data_dir)
+    dataset = tf.data.Dataset.from_tensor_slices(filenames)
+
+    # Convert to individual records
+    dataset = dataset.interleave(tf.data.TFRecordDataset,
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/image/imagenet_dataset.py:92]: does this support distributed file system, such as HDFS and S3?

File: pyzoo/zoo/examples/tensorflow/tfpark/inception/inception.py [hunk 2]
```diff
@@ -14,34 +14,246 @@
 # limitations under the License.
 #
 
-from math import ceil
```
Comment (jason-dai) [inline pyzoo/zoo/examples/tensorflow/tfpark/inception/inception.py:17]: move this file to `examples/orca/tf/inception`

File: pyzoo/zoo/examples/tensorflow/tfpark/inception/inception.py [hunk 3]
```diff
@@ -65,56 +277,38 @@ def config_option_parser():
...
-                                         features=(tf.float32, [224, 224, 3]),
-                                         labels=(tf.int32, [1]),
-                                         batch_size=options.batchSize,
-                                         validation_dataset=val_data)
-
-    images, labels = dataset.tensors
-
-    # As sequence file's label is one-based, so labels need to subtract 1.
-    zero_based_label = labels - 1
+    write_imagenet(raw_data=options.folder, output_path=options.imagenet)
```
Comment (jason-dai) [inline pyzoo/zoo/examples/tensorflow/tfpark/inception/inception.py:299]: change to `write_tfrecord(format="imagenet", ...)`

File: pyzoo/zoo/examples/tensorflow/tfpark/inception/inception.py [hunk 4]
```diff
@@ -14,34 +14,246 @@
...
+
+    dataset = dataset.map(lambda value: parse_record_fn(value, is_training, dtype))
+
+    dataset = dataset.prefetch(buffer_size=tf.data.experimental.AUTOTUNE)
+
+    return dataset
+
+
+def input_fn(is_training, data_dir):
+    dataset = read_imagenet(data_dir, is_training)
```
Comment (jason-dai) [inline pyzoo/zoo/examples/tensorflow/tfpark/inception/inception.py:220]: change to `read_tfrecord(format="imagenet", ...)`

File: pyzoo/zoo/orca/data/image/imagenet_dataset.py (hunk 1 above)
Comment (Le-Zheng) [reply pyzoo/zoo/orca/data/image/imagenet_dataset.py]: > does this support distributed file system, such as HDFS and S3?

Yes, it supports distributed file system. For HDFS, it requires setting environment variables LD_LIBRARY_PATH (path to libhdfs.so). The environment setting requirements is updated in README and doc.

File: pyzoo/zoo/examples/orca/learn/tf/inception/inception.py [hunk 5]
```diff
@@ -25,7 +25,7 @@
 from zoo.orca import init_orca_context, stop_orca_context
 from zoo.orca.learn.tf.estimator import Estimator
 from zoo.orca.learn.trigger import EveryEpoch, SeveralIteration
-from zoo.orca.data.image.imagenet_dataset import write_imagenet, read_imagenet
+from zoo.orca.data.image.tfrecord_dataset import write_tfrecord, read_tfrecord
```
Comment (jason-dai) [inline pyzoo/zoo/examples/orca/learn/tf/inception/inception.py:28]: from zoo.orca.data.image import write_tfrecord, read_tfrecord

File: pyzoo/zoo/orca/data/image/tfrecord_dataset.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    else:
+        return [
+            os.path.join(data_dir, 'validation-%05d-of-00128' % i)
+            for i in range(_NUM_IMAGENET_VAL_FILES)]
+
+
+def write_tfrecord(format, output_path, *args, **kwargs):
+    supported_format = {"imagenet"}
+    if format not in supported_format:
+        raise ValueError(format + " is not supported ")
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/image/tfrecord_dataset.py:88]: what is supported?

File: pyzoo/zoo/orca/data/image/tfrecord_dataset.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    format_to_function = {"imagenet": (write_imagenet, ["imagenet_path"])}
+    func, required_args = format_to_function[format]
+    _check_arguments(format, kwargs, required_args)
+    func(output_path=output_path, *args, **kwargs)
+
+
+def read_tfrecord(format, path, *args, **kwargs):
+    supported_format = {"imagenet"}
+    if format not in supported_format:
+        raise ValueError(format + " is not supported")
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/image/tfrecord_dataset.py:99]: what is supported?

File: pyzoo/zoo/orca/data/image/tfrecord_dataset.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        return [
+            os.path.join(data_dir, 'train-%05d-of-01024' % i)
+            for i in range(_NUM_IMAGENET_TRAIN_FILES)]
+    else:
+        return [
+            os.path.join(data_dir, 'validation-%05d-of-00128' % i)
+            for i in range(_NUM_IMAGENET_VAL_FILES)]
+
+
+def write_tfrecord(format, output_path, *args, **kwargs):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/image/tfrecord_dataset.py:85]: add python doc

File: pyzoo/zoo/orca/data/image/tfrecord_dataset.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    if format not in supported_format:
+        raise ValueError(format + " is not supported ")
+
+    format_to_function = {"imagenet": (write_imagenet, ["imagenet_path"])}
+    func, required_args = format_to_function[format]
+    _check_arguments(format, kwargs, required_args)
+    func(output_path=output_path, *args, **kwargs)
+
+
+def read_tfrecord(format, path, *args, **kwargs):
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/image/tfrecord_dataset.py:96]: add python doc

**Comment by Le-Zheng** [comment]
jenkins passed: http://10.112.231.51:18888/view/ZOO-PR/job/ZOO-PR-Validation/6218/
This pr is merged first. InceptionV1 example will move to next pr.

**Pull request merged**